### PR TITLE
CXX: Improve lambda parsing. Fix #1763.

### LINKED ISSUE #1763: C++: top level lambda
The name of the parser: C++

The command line you used to run ctags: 

```
./ctags --options=NONE --kinds-C++='*' -o - /tmp/bar.cc
```

The content of input file:

```C
auto add1i = [](auto t) {
  return 1 + t;
};
```

The tags output you are not satisfied with:

ctags prints nothing.
```
```

The tags output you expect:

```
add1i ...
```

The version of ctags:

```
$ ./ctags --version
Universal Ctags 0.0.0(44a0e979), Copyright (C) 2015 Universal Ctags Team
Universal Ctags is derived from Exuberant Ctags.
Exuberant Ctags 5.8, Copyright (C) 1996-2009 Darren Hiebert
  Compiled: Jun  9 2018, 04:19:23
  URL: https://ctags.io/
  Optional compiled features: +wildcards, +regex, +iconv, +debug, +option-directory, +xpath, +json, +interactive, +sandbox, +yaml, +aspell
```

How do you get ctags binary:

Self building

### PR DESCRIPTION
**Pull request by pragmaware**
Improve lambda parsing. Fix #1763.

### CODE REVIEW

File: parsers/cxx/cxx_parser_block.c [hunk 1]
```diff
@@ -121,6 +121,25 @@ bool cxxParserParseBlockHandleOpeningBracket(void)
...
+		bIsCPP &&
+		(pParenthesis = cxxParserOpeningBracketIsLambda())
+	)
+	{
+		if(!cxxParserHandleLambda(pParenthesis))
+		{
+			CXX_DEBUG_LEAVE_TEXT("Lambda handling failed");
+			return false;
+		}
+
```
Comment (masatake) [inline parsers/cxx/cxx_parser_block.c]: Please, remove trailing whitespace.

trailing-whitespace

**Comment by pragmaware** [comment]
Done.

By the way: it's very easy to make some empty lines with spaces slip in the commit. Why trailing whitespace is considered so bad? Can't we just run a "remove trailing whitespace" script once in a while?

File: parsers/cxx/cxx_parser_block.c (hunk 1 above)
Comment (pragmaware) [reply parsers/cxx/cxx_parser_block.c]: Done

**Comment by masatake** [comment]
> By the way: it's very easy to make some empty lines with spaces slip in the commit. Why trailing whitespace is considered so bad? Can't we just run a "remove trailing whitespace" script once in a while?

Two mutual recursive questions are included in your question: "Why it is bad" and "How to remove it".

Why:
"trailing whitespace" violates a principle that a commit only includes "necessary and sufficient" changes.
git diff renders "trailing whitespace"  red. I guess it warns a change includes unnecessary things.
(I'm talking about new code.)

How:
As you wrote, we can remove all trailing whitespaces with a script. It affects only already committed
code. It doesn't help us keep the principle in a new commit.

We should run such a script before making a pull request. It helps us keep the principle in new code.
It seems that git has such feature. However, I don't introduce it to our .git/config because (1) expected.tags needs "trailing whitespace" for comparison, (2) I think we have much chance to run git diff and git show before making a pull request. git renders "trailing whitespace" red, so we also have much change to remove it.

**Comment by masatake** [comment]
...You made two commits in this pull request.

A. One is fix the original bug I reported. It includes one trailling whitespace.
B. Another ons removes trailling whitespaces. It removes trailling whitespaces already existing and the trailling whitespace you introduced in the last commit A.

I want you to make one commit.
A'. fixing the original bug I reported without introducing new trailling whitespace.

Following are acceptable:
A''. Removing trailling whitespaces already existing
B'. fixing the original bug I reported without introducing new trailling whitespace.

**Comment by pragmaware** [comment]
Done.

**Comment by masatake** [comment]
Thank you.

**Pull request merged**